### PR TITLE
IPv6 system tests: Enrollment

### PR DESCRIPTION
**Pull request by juliamagan**
|Related issue|
|---|
| #2266  |

## Description

We have added a system test that checks if the agent enrollment works correctly for IPv4 and IPv6. This test changes the manager IP in the agent's `ossec.conf` and restarts the service. Finally, it checks the manager and agent `ossec.log`, `client.keys`, and if the agent is `Active`. 

### Network configuration

| Manager |  Agent 	|  
|:--:|:--: |
| IPV4 	| IPV4|  	
| IPV6 	| IPV6 	| 
| Both	| Both 	| 
| Both	| IPV6 	|
| Both 	| IPV4 	| 
| IPV6| Both 	|
| IPV4| Both 	|

## Testing

With wazuh/wazuh `master` branch:

| Linux | Result | Notes |
|:--:|:--:|:--:|
|R1| 🔴 | We could not deploy the environment because the `ipv6` tag is invalid |

## Checks

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the [QA-Docs Schema 2.0](https://github.com/wazuh/wazuh-qa/wiki/QA-Documentation---How-to-document-a-test-using-Schema-2.0).

### CODE REVIEW

**Comment by juliamagan** [comment]
## 15/12/2021


### Testing

We are testing in wazuh/wazuh `dev-support-ipv6-fix` development branch:

|Linux|Results|
|:--:|:--:|
|R1| [🔴 ](https://github.com/wazuh/wazuh-qa/files/7736785/R1-enrollment.zip)| 

We got some timeout errors but we couldn't check them because the logs aren't stored.

## 17/12/2021


### Testing

Now, we can check the logs received in `ossec.log`. 

We are testing in wazuh/wazuh `dev-support-ipv6-fix` development branch:

|Linux|Results|
|:--:|:--:|
|R1| 🔵 |

**Comment by juliamagan** [comment]
## 21/12/2021

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6-fix` development branch:


|Linux|Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/7755673/R1-test-enrollment.zip)

**Comment by juliamagan** [comment]
## 27/12/2021

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6` development branch:


|Linux|Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/7780981/R1-test-enrollment.zip)
|R2| 🔵 
|R3|

**Comment by juliamagan** [comment]
## 28/12/2021

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6` development branch:


|Linux|Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/7786575/R1-test-enrollment.zip)
|R2| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7786576/R2-test-enrollment.zip)
|R3| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7786578/R3-test-enrollment.zip)

File: tests/system/test_enrollment/test_enrollment.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        elif 'ipv6' in configuration['ip_type']:
+            messages_with_ip = messages_with_ip.replace('AGENT_IP', f"{network['agent_network'][1]}")
+        elif 'dns' in configuration['ip_type']:
+            if 'yes' in configuration['ipv6_enabled']:
+                if 'ipv4' in configuration['agent_network']:
+                    messages_with_ip = messages_with_ip.replace('AGENT_IP', f"{network['agent_network'][0]}")
+                else:
+                    messages_with_ip = messages_with_ip.replace('AGENT_IP', f"{network['agent_network'][1]}")
+            else:
+                messages_with_ip = messages_with_ip.replace('AGENT_IP', f"{network['agent_network'][0]}")
```
Comment (Rebits) [inline tests/system/test_enrollment/test_enrollment.py:171]:
```suggestion
        address_ip = ''
        message_ip_manager = ''
        message_ip_agent = ''
        final_message = ''

        if configuration['ip_type'] == 'ipv4':
            address_ip = network['manager_network'][0]
            message_ip_manager = address_ip
            message_ip_agent = address_ip
        elif configuration['ip_type'] == 'ipv6':
            address_ip = network['manager_network'][1]
            message_ip_manager = address_ip
            message_ip_agent = address_ip
        else:
            address_ip = 'wazuh-manager'
            if not (configuration['manager_network'] == 'ipv4' or configuration['agent_network'] == 'ipv4'):
                message_ip_manager = network['manager_network'][1]
            else:
                message_ip_manager =  network['manager_network'][0]

            if configuration['agent_network'] != 'ipv6':
                message_ip_agent = network['agent_network'][0]
            else:
                message_ip_agent = network['agent_network'][0]



        new_configuration = old_agent_configuration.replace('<address>MANAGER_IP</address>',
                                                            f"<address>{address_ip}</address>")
        host_manager.modify_file_content(host='wazuh-agent1', path='/var/ossec/etc/ossec.conf',
                                        content=new_configuration)

        final_message = messages.replace('MANAGER_IP', message_ip_manager)
        final_message = messages.replace('AGENT_IP', message_ip_agent)
```

File: tests/system/test_enrollment/test_enrollment.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+    with open(manager_conf_file, 'w') as file:
+        file.write(new_manager_configuration)
```
Comment (Rebits) [inline tests/system/test_enrollment/test_enrollment.py:174]: Use `write_file` function of the file module instead.

File: tests/system/test_enrollment/test_enrollment.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+    host_manager.apply_config(manager_conf_file)
+
+    with open(messages_path, 'w') as file:
+        file.write(messages_with_ip)
```
Comment (Rebits) [inline tests/system/test_enrollment/test_enrollment.py:179]: Same

File: tests/system/test_enrollment/test_enrollment.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+    yield
+
+    with open(messages_path, 'w') as file:
+        file.write(messages)
+
+    with open(manager_conf_file, 'w') as file:
+        file.write(old_manager_configuration)
```
Comment (Rebits) [inline tests/system/test_enrollment/test_enrollment.py:187]: Same

File: tests/system/test_enrollment/test_enrollment.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        - '.*Agent key generated for.*'
+        - '.*Server IP Address: MANAGER_IP'
+        - '.*Requesting a key from server: MANAGER_IP'
+        - '.*Registering agent to unverified manager'
+        - '.*Using agent name as:*'
+        - '.*Waiting for server reply'
+        - '.*Valid key received'
+        - '.*Waiting .* seconds before server connection'
+    tags:
+        - authd
```
Comment (Rebits) [inline tests/system/test_enrollment/test_enrollment.py:230]: In agent part there are more modules/daemons implicated.

**Review by Rebits — CHANGES_REQUESTED** [review]
Great job, but some changes are required

File: tests/system/test_enrollment/data/test_enrollment_cases.yml [hunk 6]
```diff
@@ -0,0 +1,161 @@
+-
+  name: 'manager_both_agent_both-ipv4-yes'
+  test_case:
+  -
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/data/test_enrollment_cases.yml]: Done in 569716d

File: tests/system/test_enrollment/test_enrollment.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    with open(agent_conf_file, 'r') as file:
+        old_agent_configuration = file.read()
+
+    with open(messages_path, 'r') as file:
+        messages = file.read()
+
+    with open(manager_conf_file, 'r') as file:
+        old_manager_configuration = file.read()
+
+    for configuration in test_case['test_case']:
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 569716d

File: tests/system/test_enrollment/test_enrollment.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    yield
+
+    with open(messages_path, 'w') as file:
+        file.write(messages)
+
+    with open(manager_conf_file, 'w') as file:
+        file.write(old_manager_configuration)
+
+
+@pytest.mark.parametrize('test_case', [cases for cases in test_cases_yaml], ids=[cases['name']
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 569716d

File: deps/wazuh_testing/wazuh_testing/tools/system.py [hunk 9]
```diff
@@ -244,3 +244,12 @@ def run_shell(self, host: str, cmd: str, check: bool = False):
...
         return self.get_host(host).ansible('shell', cmd, check=check)['stdout']
+
+    def get_host_ip(self, host: str):
+        """Get the Ansible object for communicating with the specified host.
+        Args:
+            host (str): Hostname
+        Returns:
+            testinfra.modules.base.Ansible: Host instance from hostspec
+        """
+        return self.get_host(host).interface('eth0').addresses
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply deps/wazuh_testing/wazuh_testing/tools/system.py]: Done in dd8fdaf

File: tests/system/provisioning/basic_environment/ansible.cfg [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+gather_timeout = 300
+stdout_callback = yaml
+callback_whitelist = profile_tasks, timer, mail
+timeout = 60
+log_path = ./ansible.log
+
+[ssh_connection]
+ssh_args = -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
+pipelining = True
+retries = 10
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/provisioning/basic_environment/ansible.cfg]: Done in 72fa2ac

File: tests/system/provisioning/basic_environment/vars/main.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+---
+include_vars: "configurations.yml"
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/provisioning/basic_environment/vars/main.yml]: Done in 72fa2ac

File: tests/system/test_enrollment/test_enrollment.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+import pytest
+
+from wazuh_testing.tools import WAZUH_PATH, WAZUH_LOGS_PATH
+from wazuh_testing.tools.file import read_yaml
+from wazuh_testing.tools.monitoring import HostMonitor
+from wazuh_testing.tools.system import HostManager
+
+
+# Hosts
+testinfra_hosts = ["wazuh-manager", "wazuh-agent1"]
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 72fa2ac

File: tests/system/test_enrollment/test_enrollment.py (hunk 4 above)
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 861bbd0

File: tests/system/test_enrollment/test_enrollment.py (hunk 3 above)
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 861bbd0

File: tests/system/test_enrollment/test_enrollment.py (hunk 2 above)
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 861bbd0

File: tests/system/test_enrollment/test_enrollment.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+def modify_ip_address_conf(test_case):
+
+    with open(agent_conf_file, 'r') as file:
+        old_agent_configuration = file.read()
+
+    with open(messages_path, 'r') as file:
+        messages = file.read()
+
+    with open(manager_conf_file, 'r') as file:
+        old_manager_configuration = file.read()
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 861bbd0

File: tests/system/test_enrollment/test_enrollment.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+
+    for configuration in test_case['test_case']:
+        if 'yes' in configuration['ipv6_enabled']:
+            new_manager_configuration = old_manager_configuration.replace('IPV6_ENABLED', "'yes'")
+        else:
+            new_manager_configuration = old_manager_configuration.replace('IPV6_ENABLED', "'no'")
+
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in cec2b59

File: tests/system/test_enrollment/test_enrollment.py (hunk 1 above)
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 641cb3c

File: tests/system/test_enrollment/test_enrollment.py (hunk 5 above)
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py:208]: Done in 8243963

File: tests/system/test_enrollment/test_enrollment.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+wait_agent_start = 20
+network = {}
+
+
+# Remove the agent once the test has finished
+@pytest.fixture(scope='function')
+def clean_environment():
+    yield
+    host_manager.control_service(host='wazuh-agent1', service='wazuh', state="stopped")
+    agent_id = host_manager.run_command('wazuh-manager', f'cut -c 1-3 {WAZUH_PATH}/etc/client.keys')
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply tests/system/test_enrollment/test_enrollment.py]: Done in 15da55c

**Comment by juliamagan** [comment]
## 14/01/2022

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6` development branch:


|Linux|Results|
|:--:|:--:|
|R1|
|R2| 
|R3|

**Comment by juliamagan** [comment]
## 18/01/2022

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6` development branch:


|Linux|Results|
|:--:|:--:|
|R1|[🔴 ](https://github.com/wazuh/wazuh-qa/files/7887961/R1-test-enrollment.zip)
|R2| [🔴 ](https://github.com/wazuh/wazuh-qa/files/7887977/R2-test-enrollment.zip)
|R3| [🔴 ](https://github.com/wazuh/wazuh-qa/files/7887979/R3-test-enrollment.zip)
 
These tests are failing because, since https://github.com/wazuh/wazuh/commit/10462c8ff36846ca4c29f76c05568dee054c7123, the IPs are expanded.

**Comment by juliamagan** [comment]
## 31/01/2022

### Testing

We are testing in wazuh/wazuh `dev-support-ipv6` development branch:


|Linux|Results|
|:--:|:--:|
|R1|[🔴 ](https://github.com/wazuh/wazuh-qa/files/7972460/R1-test-enrollment.zip)
|R2| [🔴 ](https://github.com/wazuh/wazuh-qa/files/7972462/R2-test-enrollment.zip)
|R3| [🔴 ](https://github.com/wazuh/wazuh-qa/files/7972463/R3-test-enrollment.zip)

**Pull request merged**